### ISSUE TITLE
Automatically include info about the environment in each rendered notebook

### ISSUE THREAD

**Issue by ceball** [pyviz-dev/nbsite#29]
E.g. could automatically add a link to the end of every notebook taking you to the output of conda env export / conda info (stored in a separate file).

Or a project could specify which modules matter, and the output of pyver for those modules could be included at the end of the notebook.

Or something like that.

I guess whatever we do should be executed by the notebook kernel itself, though, to be sure the right things are being reported.

**Comment by jbednar** [comment]
Good idea. We could do ``conda env attach`` to store the environment into the notebook's metadata.

**Comment by ceball** [comment]
I did not know about that command! You have to specify the environment name - do you also happen to know how to get that from python?

The reason I'd want to get it from python is that by the end of writing the original initial request above, I realized I'm uneasy about anything other than something executed by the python in the notebook kernel itself.

I am not anything like certain about this, but...the reason I'm uneasy is "kernelspecs". When nbsite runs, it currently passes `kernel_name='python%s'%sys.version_info[0]` to `nbconvert.ExecutePreprocessor`. This assumes you want to run the notebook with your `python3` kernel if you are using python 3, and your `python2` kernel if you are using python 2. But who knows what those kernels might actually be? In the python 3 conda env `envname`, you'd typically have something like this:
```
$ jupyter kernelspec list
Available kernels:
  python3    /path/to/conda/envs/envname/share/jupyter/kernels/python3
```

But you *could* have the situation where the 'python3' kernel is different from your current python3. To demonstrate, I set that up (contrived - but just to show it's possible):

```
$ jupyter kernelspec list
Available kernels:
  python3    /Users/cball/Library/Jupyter/kernels/python3
$ cat /Users/cball/Library/Jupyter/kernels/python3/kernel.json 
{
 "display_name": "Python 3", 
 "language": "python", 
 "argv": [
  "/Users/cball/anaconda3/envs/testing123/bin/python", 
  "-m", 
  "ipykernel_launcher", 
  "-f", 
  "{connection_file}"
 ]
```

So my python3 kernel is from the anaconda3 environment 'testing123'. But my current environment is miniconda environment 'testing456':

```
$ conda info --envs
# conda environments:
#
[...]
testing456                   *  /Users/cball/mc3/envs/testing456
```

If I run my notebooks using nbsite from this 'testing456' environment, the notebooks will actually run in the 'testing123' environment.

So, basically, nbsite guesses at the kernel to use. If it didn't guess, and the notebook itself
included a kernelspec (which unfortunately they frequently do), the kernel specified in the notebook itself would be used. (I just realized: nbsite should not specify a kernel, and should clear the notebook metadata before running, and then maybe there'd be an even better chance of running with the user's intended environment. Would have to check the nbconvert docs and test to see.)

Examples of notebooks containing kernelspecs:

<img width="990" alt="screen shot 2017-12-05 at 9 48 25 pm" src="https://user-images.githubusercontent.com/1929/33632706-64ec3b04-da06-11e7-88ae-459749fd5741.png">

```
$ jupyter nbconvert --execute examples/getting_started/1_Introduction.ipynb 
[NbConvertApp] Converting notebook examples/getting_started/1_Introduction.ipynb to html
[NbConvertApp] Executing notebook with kernel: conda-env-ds-py
Traceback (most recent call last):
  File "/Users/cball/Eunectes/mc3/envs/p4k/lib/python3.6/site-packages/jupyter_client/kernelspec.py", line 201, in get_kernel_spec
    resource_dir = d[kernel_name.lower()]
KeyError: 'conda-env-ds-py'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
[...]
jupyter_client.kernelspec.NoSuchKernel: No such kernel named conda-env-ds-py
```

Or if I try to open it with the notebook server:

<img width="1057" alt="screen shot 2017-12-05 at 10 06 49 pm" src="https://user-images.githubusercontent.com/1929/33633425-aec59f66-da08-11e7-94bf-f9461e1bae09.png">

Kernel names I encounter these days are more typically just 'python3' or 'python2', so people don't notice their presence (as they tend to have a kernel of the same name themselves):

<img width="1239" alt="screen shot 2017-12-05 at 10 09 23 pm" src="https://user-images.githubusercontent.com/1929/33633504-02f8d3aa-da09-11e7-8134-82a50ba28fac.png">

**Comment by ceball** [comment]
I guess for projects containing notebooks, we should just make it part of the project tests that there is no kernelspec in submitted notebooks - at least to avoid user confusion on opening notebooks. (Probably should also enforce no output, and no optional metadata at all.)

**Comment by jbednar** [comment]
We should do whatever can avoid those annoying kernel popups when starting Jupyter, yes.

Looks like ``conda env attach`` is about to be deprecated, though: https://github.com/conda/conda/issues/5842

**Comment by jbednar** [comment]
I think that's mostly true, but are we currently archiving the specific set of packages that were installed when the notebook is run?  If not, then I don't think we're actually capturing what's needed for full reproducibility.

**Comment by jsignell** [comment]
Fair enough we should probably start using anaconda-project lock to capture that.
